rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -277,15 +311,121 @@ typedef void (* signal_handler_t)(int);
 signal_handler_t rclcpp::SignalHandler::old_signal_handler = 0;
 #endif
 
-/// Mutex to protect g_sigint_guard_cond_handles
-static std::mutex g_sigint_guard_cond_handles_mutex;
-/// Guard conditions for interrupting the rmw implementation when the global interrupt signal fired.
-static std::map<rcl_wait_set_t *, rcl_guard_condition_t> g_sigint_guard_cond_handles;
+#if defined(_WIN32)
+#include <windows.h>
+#elif defined(__APPLE__)
+#include <dispatch/dispatch.h>
+#else  // posix
+#include <semaphore.h>
+#endif
+
+static std::atomic_bool g_wait_for_signal_is_setup = ATOMIC_VAR_INIT(false);
+#if defined(_WIN32)
+HANDLE g_signal_handler_sem;
+#elif defined(__APPLE__)
+dispatch_semaphore_t g_signal_handler_sem;
+#else  // posix
+sem_t g_signal_handler_sem;
+#endif
```

Comment:
@wjwwood I think these variables could be pushed into the `SignalHandler` singleton.

Reply:
Again, lateral change, not really an improvement in my opinion, in fact kind of worse because you have initialize them outside of the class anyways (violates D.R.Y. more than necessary).

Reply:
I did actually move them in my latest changes, however, the above is just in retrospect.